runtime/compiler/optimizer/SequentialStoreSimplifier.cpp

```diff
@@ -677,621 +684,2200 @@ TR::Node* getALoadReferenceForSeqLoad(TR::Node* rootNode, int32_t totalBytes, in
+
+         if (secondChild->getOpCodeValue() != TR::lconst)
+            {
+            return false;
+            }
+         }
+      }
+   else
       {
-      dumpOptDetails(_comp, " Remove trees %p to %p\n", _treeTops[i]->getNode(), _treeTops[i]->getNextTreeTop()->getNode());
-      TR::TreeTop::removeDeadTrees(_comp, _treeTops[i], _treeTops[i]->getNextTreeTop());
+      /*
+       * Under 32bit, bloadi's child must be an aiadd node. Some of the aiadd nodes go away after the transformation.
+       * So, if the refCount is not 1, bail out here.
+       */
+      if ((firstChild->getOpCodeValue() != TR::aiadd) || (firstChild->getReferenceCount() != 1))
+         {
+         return false;
+         }
+
+      secondChild = firstChild->getSecondChild();
+      firstChild = firstChild->getFirstChild();
+
+      /* aiadd's first child can be aload or aloadi. This load is kept so there is no refcount restriction. */
+      if ((firstChild->getOpCodeValue() != TR::aload) && (firstChild->getOpCodeValue() != TR::aloadi))
+         {
+         return false;
+         }
+
+      /*
+       * aiadd's second child can be an iconst, iadd or isub. The iadd or isub might go away after transformation
+       * so they must have a refCount of 1.
+       */
+      if (secondChild->getReferenceCount() != 1)
+         {
+         if (secondChild->getOpCodeValue() != TR::iconst)
+            {
+            return false;
+            }
+         }
+      else
+         {
+         if ((secondChild->getOpCodeValue() != TR::iconst) &&
+             (secondChild->getOpCodeValue() != TR::iadd)   &&
+             (secondChild->getOpCodeValue() != TR::isub))
+            {
+            return false;
+            }
+         }
+
+      /* If the second child is isub or iadd, make sure that node's second child is an iconst. */
+      if (secondChild->getOpCodeValue() != TR::iconst)
+         {
+         secondChild = secondChild->getSecondChild();
+
+         if (secondChild->getOpCodeValue() != TR::iconst)
+            {
+            return false;
+            }
+         }
       }
 
+   return true;
    }
 
-int64_t TR_arraycopySequentialStores::constVal()
+int64_t getOffsetForSeqLoad(TR::Compilation* comp, TR::Node* byteConversionNode)
    {
-   int dir;
-   int baseShift;
-   uint64_t val = 0;
+   /*
+    * Returns the constant offset from basePointerNode. Offset from aload/aloadi if there is no basePointerNode.
+    * byteConversionNode must be: b2i, b2l, bu2i, bu2l
+    *
+    * Example tree: (sub/add version)
+    * b2i        <--- byteConversionNode
+    *   bloadi
+    *     aladd
+    *       aload/aloadi
+    *       lsub/ladd     <<--- displacementNode
+    *         **basePointerNode**
+    *         lconst      <<--- return value
+    *
+    *
+    * Example tree: (const version)
+    * b2i        <--- byteConversionNode
+    *   bloadi
+    *     aladd
+    *       aload/aloadi
+    *       lconst        <<--- displacementNode, return value
```

Comment:
Do we allow byte loads without the basePointerNode? How would we match then nodes that have it and the ones that don't?

Reply:
Yes. This happens if the index into the array are constants like `a[0], a[1], a[2], a[3]`. In this case `basePointerNode` is treated as being NULL since it does not exist. If it is NULL in one of the byte loads that are being combined it must be NULL for the others as well to be a match. If it is NULL for one load and not for another, it is not considered a match.